rust/opendp/src/any.rs

```diff
@@ -0,0 +1,393 @@
+    move |d_in: &AQI, d_out: &AQO| {
+        let d_in = d_in.downcast_ref()?;
+        let d_out = d_out.downcast_ref()?;
+        relation(d_in, d_out)
+    }
+}
+
+fn make_any_map<QI, QO, AQI>(map: &Option<Rc<dyn Fn(&QI) -> Fallible<Box<QO>>>>) -> Option<impl Fn(&AQI) -> Fallible<Box<AnyMetricDistance>>>
+    where QI: 'static + PartialOrd,
+          QO: 'static + PartialOrd + Clone,
+          AQI: Downcast {
+    map.as_ref().map(|map| {
+        let map = map.clone();
+        move |d_in: &AQI| -> Fallible<Box<AnyMetricDistance>> {
+            let d_in = d_in.downcast_ref()?;
+            let d_out = map(d_in);
+            d_out.map(|d| AnyMetricDistance::new(*d)).map(Box::new)
+        }
+    })
+}
+
+impl<MI: Metric, MO: Measure> PrivacyRelation<MI, MO>
+    where MI::Distance: 'static + Clone + PartialOrd,
+          MO::Distance: 'static + Clone + PartialOrd {
+    pub fn as_any(&self) -> PrivacyRelation<AnyMetric, AnyMeasure> {
+        PrivacyRelation::new_all(
+            make_any_relation(&self.relation),
+            make_any_map(&self.backward_map),
+        )
+    }
+}
+
+impl<MI: Metric, MO: Metric> StabilityRelation<MI, MO>
+    where MI::Distance: 'static + Clone + PartialOrd,
+          MO::Distance: 'static + Clone + PartialOrd {
+    pub fn as_any(&self) -> StabilityRelation<AnyMetric, AnyMetric> {
+        StabilityRelation::new_all(
+            make_any_relation(&self.relation),
+            make_any_map(&self.forward_map),
+            make_any_map(&self.backward_map),
+        )
+    }
+}
+
+impl<DI: 'static + Domain, DO: 'static + Domain, MI: 'static + Metric, MO: 'static + Measure> Measurement<DI, DO, MI, MO>
+    where DI::Carrier: 'static,
+          DO::Carrier: 'static,
+          MI::Distance: 'static + Clone + PartialOrd,
+          MO::Distance: 'static + Clone + PartialOrd {
+    pub fn into_any(self) -> Measurement<AnyDomain, AnyDomain, AnyMetric, AnyMeasure> {
+        Measurement::new(
+            AnyDomain::new(*self.input_domain),
+            AnyDomain::new(*self.output_domain),
+            self.function.as_any(),
+            AnyMetric::new(*self.input_metric),
+            AnyMeasure::new(*self.output_measure),
+            self.privacy_relation.as_any(),
+        )
+    }
+}
+
+impl<DI: 'static + Domain, DO: 'static + Domain, MI: 'static + Metric, MO: 'static + Metric> Transformation<DI, DO, MI, MO>
+    where DI::Carrier: 'static,
+          DO::Carrier: 'static,
+          MI::Distance: 'static + Clone + PartialOrd,
+          MO::Distance: 'static + Clone + PartialOrd {
+    pub fn into_any(self) -> Transformation<AnyDomain, AnyDomain, AnyMetric, AnyMetric> {
+        Transformation::new(
+            AnyDomain::new(*self.input_domain),
+            AnyDomain::new(*self.output_domain),
+            self.function.as_any(),
+            AnyMetric::new(*self.input_metric),
+            AnyMetric::new(*self.output_metric),
+            self.stability_relation.as_any(),
+        )
+    }
+}
```

Comment:
I was surprised that any.rs is a member of the opendp crate, as it seems like an FFI feature. But you "can't define an inherent impl" for a foreign type. If these were split off into standalone functions, then this could be moved outside of core.

```rust
pub fn transformation_into_any<DI: 'static + Domain, DO: 'static + Domain, MI: 'static + Metric, MO: 'static + Metric>(
    trans: Transformation<DI, DO, MI, MO>
) -> Transformation<AnyDomain, AnyDomain, AnyMetric, AnyMetric>
    where DI::Carrier: 'static,
          DO::Carrier: 'static,
          MI::Distance: 'static + Clone + PartialOrd,
          MO::Distance: 'static + Clone + PartialOrd {
    Transformation::new(
        AnyDomain::new(*trans.input_domain),
        AnyDomain::new(*trans.output_domain),
        trans.function.as_any(),
        AnyMetric::new(*trans.input_metric),
        AnyMetric::new(*trans.output_metric),
        trans.stability_relation.as_any(),
    )
}
```

I see it as, any.rs is in core for the .to_any method convenience.

Reply:
I take it back- while this type isn't explicitly used in interactive.rs, in practice you're likely to want to use it to erase the type from the interactive composition queryable.

Reply:
That's a great point. I think your initial instinct was right, `any.rs` should be in opendp-ffi. For the more limited form of type erasure we need for composition, I made `poly.rs`. It just didn't occur to me to move the any stuff to opendp-ffi, but I *think* it can live there.